bdk_electrum/src/lib.rs

```diff
@@ -0,0 +1,800 @@
+                }
+            }
+
+            if !params.outpoints.is_empty() {
+                match self.populate_with_outpoints(
+                    &to_position,
+                    &mut update,
+                    params.outpoints.iter().cloned(),
+                ) {
+                    Err(InternalError::Reorg) => continue,
+                    Err(InternalError::ElectrumError(e)) => return Err(e.into()),
+                    Ok(_txs) => { /* [TODO] cache full txs to reduce bandwidth */ }
+                }
+            }
+
+            // check for reorgs during scan process
+            let our_tip = update
+                .latest_checkpoint()
+                .expect("update must have atleast one checkpoint");
+            let server_blockhash = self.block_header(our_tip.height as usize)?.block_hash();
+            if our_tip.hash != server_blockhash {
+                continue; // reorg
+            } else {
+                break update;
+            }
+        };
+
+        let last_active_index = request_spks
+            .into_keys()
+            .filter_map(|k| {
+                scanned_spks
+                    .range((k.clone(), u32::MIN)..=(k.clone(), u32::MAX))
+                    .rev()
+                    .find(|(_, (_, active))| *active)
+                    .map(|((_, i), _)| (k, *i))
+            })
+            .collect::<BTreeMap<_, _>>();
+
+        Ok(ElectrumUpdate {
+            update,
+            last_active_indices: last_active_index,
+        })
+    }
+
+    /// Convenience method to call [`scan`] without requiring a keychain.
+    ///
+    /// [`scan`]: ElectrumClient::scan
+    pub fn scan_without_keychain<P, S>(
+        &self,
+        to_position: impl Fn(ChainPositionArgs<'_>) -> Result<P, Error>,
+        local_chain: &BTreeMap<u32, BlockHash>,
+        params: ScanParamsWithoutKeychain<S>,
+    ) -> Result<ElectrumUpdate<(), P>, Error>
+    where
+        P: ChainPosition,
+        S: IntoIterator<Item = Script>,
+    {
+        let spk_iter = params
+            .spks
+            .into_iter()
+            .enumerate()
+            .map(|(i, spk)| (i as u32, spk));
+        let params = ScanParams {
+            keychain_spks: [((), spk_iter)].into(),
+            txids: params.txids,
+            txs: params.txs,
+            outpoints: params.outpoints,
+            stop_gap: params.stop_gap,
+            batch_size: params.batch_size,
+        };
+        self.scan(to_position, local_chain, params)
+    }
+}
+
+/// Parameters for [`ElectrumClient::scan`].
+pub struct ScanParams<K, S: IntoIterator<Item = (u32, Script)>> {
+    /// Indexed script pubkeys of each keychain to scan transaction histories for.
+    pub keychain_spks: BTreeMap<K, S>,
+    /// Txids to scan for. The update will update the [`ChainPosition`]s for these.
+    pub txids: Vec<Txid>,
+    /// Full transactions to scan for. The update will update the [`ChainPosition`]s for these.
+    pub txs: Vec<Transaction>,
```

Comment:
I see this is here because it's faster to look up if we already have the tx. Can we not though. It's a super weird API!